Source/KeyHandler.mm

```diff
@@ -1860,8 +1860,9 @@ - (InputStateChoosingCandidate *)_buildCandidateState:(InputStateNotEmpty *)curr
         InputStateCandidate *candidate = [[InputStateCandidate alloc] initWithReading:r value:v displayText:dt];
         [candidatesArray addObject:candidate];
     }
+    InputStateInputting *inputting = (InputStateInputting *)[self buildInputtingState];
 
-    InputStateChoosingCandidate *state = [[InputStateChoosingCandidate alloc] initWithComposingBuffer:currentState.composingBuffer cursorIndex:_grid->cursor() candidates:candidatesArray useVerticalMode:useVerticalMode];
+    InputStateChoosingCandidate *state = [[InputStateChoosingCandidate alloc] initWithComposingBuffer:inputting.composingBuffer cursorIndex:inputting.cursorIndex candidates:candidatesArray useVerticalMode:useVerticalMode];
```

Comment:
Let's take a step back. The change to use `_grid->cursor()` (and hence the issue) was introduced in https://github.com/openvanilla/McBopomofo/pull/424/commits/e3b559594f8556cd13587a4805533f1c363adb6e because we mutated the grid's cursor to accomodate the new logic with the MS IME cursor mode. Could we do one of the following, either:

(1) Make it clear that the method `_buildCandidateState:useVerticalMode:` will always disregard the current state the IME is in, and will always build an inputting state on its own? If you take this path, you might as well removed the `currenState` parameter from this method. Or,

(2) Rebuild an inputting state after the cursor is mutated ([here](https://github.com/openvanilla/McBopomofo/blob/ae4146f3246ef230727834d163a8c518d686c4d1/Source/KeyHandler.mm#L545) and [here](https://github.com/openvanilla/McBopomofo/blob/ae4146f3246ef230727834d163a8c518d686c4d1/Source/KeyHandler.mm#L661), and revert this argument to `currentState.cursorIndex`?

I personally perfer (2)–I believe we wrote `_buildCandidateState:useVerticalMode:` with the intent that it returns a ChoosingCandidate state that _inherits_ the properties from `currentState`. Rebuilding an inputting state after cursor mutation seems reasonable, since the mutation sites have the knowledge that they must be in an input state. I try to avoid embedding too many implicit knowledge in an already complex state machinery. It should also lead to less code change than approach (1), but I'll defer to your judgement.